dandiapi/api/tests/factories.py

```diff
@@ -200,38 +223,45 @@ def metadata(self):
 
 
 class PublishedAssetFactory(DraftAssetFactory):
+    published = True
+    status = Asset.Status.VALID  # published assets are always valid
+
     @classmethod
-    def _create(cls, *args, **kwargs):
-        asset: Asset = super()._create(*args, **kwargs)
-        asset.status = Asset.Status.VALID  # published assets are always valid
+    def _create(cls, model_class: type[Asset], *args, **kwargs) -> Asset:
+        # Call "_build" to create without saving, as a save won't be valid until the metadata is set
+        asset: Asset = super()._build(model_class, *args, **kwargs)
+        asset.metadata = asset.published_metadata()
         asset.save()
-        publish_asset(asset=asset)
-        asset.refresh_from_db()
         return asset
 
 
 class UploadFactory(factory.django.DjangoModelFactory):
     class Meta:
         model = Upload
+        skip_postgeneration_save = True
 
     upload_id = factory.Faker('uuid4')
     multipart_upload_id = factory.Faker('uuid4')
-    blob = factory.django.FileField(data=factory.Faker('binary', length=100))
+    blob = factory.django.FileField(
+        filename=factory.Faker('file_name', extension='dat'),
+        data=factory.Faker('binary', length=factory.SelfAttribute('...size')),
+    )
+    size = 100
```

Comment:
I'm actually not sure the ideal way to structure this. I can optimize this to set a user-defined `size`, whereby it will auto-generate the `blob`:
```python
UploadFactory.create(size=...)
```
Or I can optimize this to set some content, whereby it will auto-populate the `size`:
```python
UploadFactory.create(blob__data=b'abcd')
```

I don't think I can easily support both. Which is a more useful pattern?

Reply:
I'd tend to think a user-defined `size` field would be more commonly used, although I think they'd both be used.